### ISSUE TITLE
Optimize button

### ISSUE THREAD

**Issue by grumpyinca** [thegrumpys/odop#613]
The **Search** button introduced in ODOP v4.0 resulted in a significant increase in user engagement and general level of success with the software.  The Search button appears only when a design is not feasible.

This issue proposes an **Optimize** button to appear in the same location as the Search button but only after completion of a Search that produces a feasible design.  It is debatable whether the Optimize button should appear for the default designs loaded on initial program startup.  

In its simplest form, the Optimize button could have the same action as the Action : Seek menu.  This level of function is the highest priority and the only aspect referred to by the "Hot" label.

Alternatively, the Optimize button could bring up a modal dialog that includes multiple elements:
 - Introductory text.  For example: 
 `"If one feasible design exists there are likely many more available, each with varying advantages / disadvantages.  The ODOP Seek feature provides a “goal seeking” capability to optimize your design on the parameter that you specify.  If starting with a default design, additional constraints specific to your application are required to obtain meaningful results.  Click Help (below) for details."`
 - Design type specific shortcut buttons.  For example: "Seek MIN Weight", "Seek MAX Cycle_Life", "Seek MIN Rate", "Seek MIN L_Solid"
 - A  "**Seek Specific**" button that has same action as the Action : Seek menu.
 - A "**Seek All**" button (see below)
 - A context "Help" button
 - A Close / Done / Cancel button
 
The Seek All button would transfer to a new modal dialog that would have its own introductory text.  This could include words similar to:
`"Press the Seek All button below to calculate the constrained MIN and MAX values for all variables not in Fixed status.  Note that this computation could take several minutes to complete the entire table"`

The table produced might look something like:
Variable Name| &nbsp; | MIN|MAX
--- | --- | --- | --- 
IV1|   | x.xx | y.yy  
IV2|   | x.xx | y.yy  
IV3|   | x.xx | y.yy  
DV1|   | x.xx | y.yy  
DV2|   | x.xx | y.yy  
DV3|   | x.xx | y.yy  
DV4|   | x.xx | y.yy  

The variables in this table would be the same Fixed-variables-excluded list used by Action : Seek.  Ideally, the individual MAX & MIN values would fill in as they become available.  This would serve as a progress indicator.  Perhaps the cell for the value actively being computed could get a colored border, icon or animation.  

Note that there likely needs to be provisions for failed or degenerate results.  For example: 
 - The case where Seek fails to improve its starting values
 - The case where Action : Seek returns a message to the effect of "Run Seek again to improve results".  
 - Sometimes Seek returns results that are technically not feasible.  Perhaps the final cell border should get a different color depending on the associated resulting objective value.

The Seek All modal footer buttons would be "Seek All", "Cancel" and context "Help".  A "Stop" button may be desirable to allow the user to end the calculation early.

See #615.  It is possible that work on both of these issues should be coordinated.

**Comment by grumpyinca** [comment]
In order to better understand the implications of "Seek All", I manually went through the process and transcribed the results below.

The following constraints were added to the default set in the compression spring Startup design:
Force_2 MIN 40.1
Deflect_2 MIN 1.75
Spring_Index MIN 4.0
Spring_Index MAX 25

Search was run after these constraints were applied and the resulting design was saved into the design library.  This design was then reestablished prior to each Seek operation.  In roughly 20-30% of the Seeks, the Result Section message was "Run Seek again to improve results".  In this situation, Seek was repeated at least once and on a few occasions, more than twice.

Variable Name| &nbsp; | Start Value | &nbsp; | MIN|MAX
--- | --- | --- | --- | --- | --- 
OD_Free|   | z.zz |   | x.xx | y.yy  
Wire_Dia|   | z.zz |   | x.xx | y.yy  
L_Free|   | z.zz |   | x.xx | y.yy  
Coils_T|   | z.zz |   | x.xx | y.yy  
Force_1|   | z.zz |   | x.xx | y.yy  
Force_2|   | z.zz |   | x.xx | y.yy  
Mean_Dia|   | z.zz |   | x.xx | y.yy  
Coils_A|   | z.zz |   | x.xx | y.yy  
Rate|   | z.zz |   | x.xx | y.yy  
Deflect_1|   | z.zz |   | x.xx | y.yy  
Deflect_2|   | z.zz |   | x.xx | y.yy  
L_1|   | z.zz |   | x.xx | y.yy  
L_2|   | z.zz |   | x.xx | y.yy  
L_Stroke|   | z.zz |   | x.xx | y.yy  
Slenderness|   | z.zz |   | x.xx | y.yy  
ID_Free|   | z.zz |   | x.xx | y.yy  
Weight|   | z.zz |   | x.xx | y.yy  
Spring_Index|   | z.zz |   | x.xx | y.yy  
Force_Solid|   | z.zz |   | x.xx | y.yy  
Stress_1|   | z.zz |   | x.xx | y.yy  
Stress_2|   | z.zz |   | x.xx | y.yy  
Stress_Solid|   | z.zz |   | x.xx | y.yy  
FS_2|   | z.zz |   | x.xx | y.yy  
FS_Solid|   | z.zz |   | x.xx | y.yy  
FS_CycleLife|   | z.zz |   | x.xx | y.yy  
%_Avail_Deflect|   | z.zz |   | x.xx | y.yy  
Energy|   | z.zz |   | x.xx | y.yy

**Comment by grumpyinca** [comment]
A possible Optimize button modal:

![OptimizeButtonModal1](https://user-images.githubusercontent.com/39974784/148669037-623aec52-db5a-44a2-89f0-fd1b39412ed0.png)

**Comment by 1fifoto** [comment]
Coded as follows, an info icon to the right of Seek/Optimize provides Pop-up help.

<img width="417" alt="Screen Shot 2022-01-16 at 5 51 48 PM" src="https://user-images.githubusercontent.com/1581602/149683386-e63700f9-97cf-4b5e-9ff8-b9990c0de20b.png">

When the button is pushed the Seek/Optimize modal is displayed allowing the user to select pre-configured Seek optimizations.

<img width="1155" alt="Screen Shot 2022-01-16 at 5 55 55 PM" src="https://user-images.githubusercontent.com/1581602/149683398-fd40159b-7a74-415f-ad2c-cda582939e08.png">

**Comment by 1fifoto** [comment]
Please review the UI and the text content and provide feedback in this issue as comments.

**Comment by 1fifoto** [comment]
We have a problem with the Optimize button's modal. It is built-into Result-Set component which is problem independent code but it contains MIN/MAX buttons that are problem dependent. We need to fix that.

**Comment by 1fifoto** [comment]
Improve Seek performance. Profile code somehow to determine hot spots.

**Comment by grumpyinca** [comment]
Another concern regarding the pre-configured Seek buttons ...

It does not make sense to display a pre-configured Seek button that utilizes a variable that is currently in Fixed status.  If a fix is not trivial, it brings into question the cost / benefit of those pre-configured Seek buttons.  I expect that conversation is appropriate.

**Comment by grumpyinca** [comment]
Proposed pre-configured Seek buttons:

**Piston-Cylinder**
MAX Force
MIN Radius
MIN Pressure
Min Stress

**Solid**
MAX Volume
MAX Weight
MIN Length
MIN Width
MIN Height

**Compression**
MIN Weight
MAX Cycle_Life
MIN Rate
MIN L_Solid

**Extension**
MIN Weight
MAX Cycle_Life
MIN Rate
MAX L_Stroke

**Torsion**
MIN Weight
MAX Cycle_Life
MIN Rate
MAX Stroke

**Comment by 1fifoto** [comment]
Proposes that the ResultTable buttons be labeled "Search (solve)" and "Seek (optimize)".

Search
<img width="379" alt="Screen Shot 2022-02-19 at 7 07 00 PM" src="https://user-images.githubusercontent.com/1581602/154824477-7257eb45-7a29-4e20-a3c7-315b8edc40e7.png">

Seek
<img width="368" alt="Screen Shot 2022-02-19 at 7 07 12 PM" src="https://user-images.githubusercontent.com/1581602/154824481-2e44a900-1fef-4c75-973a-1595346fbeb8.png">

**Comment by 1fifoto** [comment]
Summary of testing in Branch 613:

1. The Help button (line 79 of ResultTable.js) should link to ...odop/Help/seek.
1.  If a button's variable is in Fixed status, that button should not appear (or possibly appear as disabled).
1. I am seeing behavior that I cannot explain with extension spring Seek MAX L_Stroke.  Is there a chance that the button is labeled MAX but it is actually seeking a MIN ?

All issues above have been repaired.

**Comment by 1fifoto** [comment]
Merged into master, closing
